app/src/main/java/org/mozilla/fenix/library/history/HistoryAdapter.kt

```diff
@@ -54,25 +52,20 @@ class HistoryAdapter(
 
     override fun onBindViewHolder(holder: HistoryListItemViewHolder, position: Int) {
         val current = getItem(position) ?: return
-        val headerForCurrentItem = timeGroupForHistoryItem(current)
+        val headerForCurrentItem = HistoryViewModel.timeGroupForHistoryItem(current)
         val isPendingDeletion = pendingDeletionIds.contains(current.visitedAt)
         var timeGroup: HistoryItemTimeGroup? = null
 
-        // Add or remove the header and position to the map depending on it's deletion status
-        if (itemsWithHeaders.containsKey(headerForCurrentItem)) {
-            if (isPendingDeletion && itemsWithHeaders[headerForCurrentItem] == position) {
-                itemsWithHeaders.remove(headerForCurrentItem)
-            } else if (isPendingDeletion && itemsWithHeaders[headerForCurrentItem] != position) {
-                // do nothing
-            } else {
-                if (position <= itemsWithHeaders[headerForCurrentItem] as Int) {
-                    itemsWithHeaders[headerForCurrentItem] = position
-                    timeGroup = headerForCurrentItem
-                }
-            }
-        } else if (!isPendingDeletion) {
-            itemsWithHeaders[headerForCurrentItem] = position
+        if (position == 0) {
```

Comment:
Had to change the way this works because the original algorithm doesn't work if the list can be partially updated.